### PR TITLE
add int32 and int64 dtypes and replace int with int32_t

### PR DESCRIPTION
**Pull request by okdshin**

### CODE REVIEW

File: include/menoh/menoh.h [hunk 1]
```diff
@@ -34,6 +34,8 @@ extern "C" {
 /*! @ingroup vpt
  */
 enum menoh_dtype_constant {
+    menoh_dtype_int32,
+    menoh_dtype_int64,
```
Comment (msakai) [inline include/menoh/menoh.h]: This changes the value of `menoh_dtype_float` and breaks binary compatibility.
It is better to add them after `menoh_dtype_float` or to specify values explicitly to all of `menoh_dtype_constant` values.

Comment (okdshin) [reply include/menoh/menoh.h]: I see. I'll change the order.

File: menoh/array.cpp [hunk 2]
```diff
@@ -2,32 +2,47 @@
 
 namespace menoh_impl {
 
-    array::array(dtype_t d, std::vector<int> const& dims, void* data_handle)
+    array::array(dtype_t d, std::vector<int32_t> const& dims, void* data_handle)
```
Comment (Hakuyume) [inline menoh/array.cpp]: Why don't you use `uint32_t`? Is there any case that `dims` is negative?

Comment (msakai) [reply menoh/array.cpp]: numpy accepts `-1` as dimension in some places (e.g. argument for `reshape`).
How about ONNX?
If ONNX also accepts `-1` in some places, then I think it makes sense to use signed integer type here too.

Comment (Hakuyume) [reply menoh/array.cpp]: If the element of `dims` takes `-1`, `int_32/64_t` makes sense.

Comment (okdshin) [reply menoh/array.cpp]: `Reshape` operator accepts `-1`. However `menoh::array` can't have -1 dim so It is reasonable to define two type of dims `std::vector<int64_t>` and `std::vector<uint64_t>` I think.
https://github.com/onnx/onnx/blob/master/docs/Operators.md#reshape

Comment (msakai) [reply menoh/array.cpp]: Of course, I understand that `menoh::array()` and `menoh:array::dims()`  do not use `-1`.
But still I advocate that: if there are some places that uses dimension `-1` and signed vector types,  using the same signed vector type for here too is better for the consistency and simplicity of API.

File: include/menoh/menoh.h (hunk 1 above)
Comment (okdshin) [reply include/menoh/menoh.h]: done.

Comment (msakai) [reply include/menoh/menoh.h]: thanks! :+1:

File: menoh/array.cpp (hunk 2 above)
Comment (msakai) [reply menoh/array.cpp]: @okdshin @Hakuyume

Comment (okdshin) [reply menoh/array.cpp]: I choose `int32_t`. It is more simple.

Comment (Hakuyume) [reply menoh/array.cpp]: I like this idea https://github.com/pfnet-research/menoh/pull/41#discussion_r203271016. Users can know whether `-1` is accepted or not from the type of arguments.

Comment (Hakuyume) [reply menoh/array.cpp]: If using both `int32/uint32` looks confusing, it is ok to use `int32`. (In that case, I will wrap them using optional types in my own wrapper)

**Comment by msakai** [comment]
Is this PR subsumed by https://github.com/pfnet-research/menoh/pull/168 ?

**Comment by okdshin** [comment]
Yes. So I close this PR. Thanks for pointing out.

**Pull request closed without merging**